crates/astria-sequencer/src/mempool/recently_included_transactions.rs

```diff
@@ -0,0 +1,424 @@
+    pub(super) fn height(&self) -> u64 {
+        self.height
+    }
+
+    pub(super) fn result(&self) -> &ExecTxResult {
+        &self.result
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, thiserror::Error)]
+pub(super) enum IncludedTransactionError {
+    #[error("Transaction ID {0} already exists in the cache")]
+    IdAlreadyExists(TransactionId),
+    #[error("Transaction insertion failed due to an internal error")]
+    InsertionInternalError,
+}
+
+/// A cache of transaction data for transactions which were recently included in a block. The cache
+/// has a maximum size of [`MAX_RECENTLY_INCLUDED_TRANSACTIONS`] transactions. Transactions are
+/// cleared from the cache after [`MAX_TIME_RECENTLY_INCLUDED_TRANSACTIONS_SECS`] seconds if the
+/// cache is not full.
+///
+/// Supports O(1) lookups by transaction ID while maintaining O(1) insertion and removal of oldest
+/// element. Worst case time complexity for cleaning stale transactions is O(n), where n is the
+/// number of transactions in the cache.
+pub(super) struct RecentlyIncludedTransactions {
+    /// Transaction IDs in chronological order (oldest first)
+    timestamped_ids: VecDeque<(TransactionId, Instant)>,
+    /// Hash map containing transaction metadata, keyed by transaction ID
+    transactions: HashMap<TransactionId, IncludedTransactionMetadata>,
+}
+
+impl RecentlyIncludedTransactions {
+    pub(super) fn new() -> Self {
+        Self {
+            timestamped_ids: VecDeque::new(),
+            transactions: HashMap::new(),
+        }
+    }
+
+    /// Looks up transaction metadata by its transaction ID, returning
+    /// `Some(IncludedTransactionMetadata)` if found, or `None` if not found.
+    pub(super) fn get_by_tx_id(
+        &self,
+        tx_id: &TransactionId,
+    ) -> Option<&IncludedTransactionMetadata> {
+        self.transactions.get(tx_id)
+    }
+
+    /// Adds new transaction data to the cache. Cleans stale transactions before attempting to add
+    /// the new transaction. If the cache is full, removes the oldest transaction until the cache
+    /// is not full. Returns an error if the transaction ID already exists in the cache.
+    pub(super) fn add_transaction(
+        &mut self,
+        tx_id: TransactionId,
+        height: u64,
+        result: ExecTxResult,
+    ) -> Result<(), IncludedTransactionError> {
+        // If the cache is full, remove the oldest transaction until the cache is not full.
+        while self.timestamped_ids.len() >= MAX_RECENTLY_INCLUDED_TRANSACTIONS {
+            let (oldest_tx_id, _) = self
+                .timestamped_ids
+                .pop_front()
+                .ok_or(IncludedTransactionError::InsertionInternalError)?;
+            self.transactions.remove(&oldest_tx_id);
+        }
+
+        // Check if the transaction ID already exists in the cache.
+        if self.transactions.contains_key(&tx_id) {
+            return Err(IncludedTransactionError::IdAlreadyExists(tx_id));
+        }
```

Comment:
I don't think we need this extra check.  Instead, we can just do something like this when inserting:
```rust
match self.transactions.insert(...) {
    None => self.timestamped_ids.push_back((tx_id, now)),
    Some(_) => warn!("..."),
}
```